File: ch01/python/lottery_fastest.py
MAX_N = 1000

# 標準入力から値を取得
n = int(input())
m = int(input())
k = []
for i in range(n):
    k.append(int(input()))

kk = []
for c in range(n):
    for d in range(n):
        kk.append(k[c] + k[d])
kk.sort()

# 二分探索
def binary_search(x):
    # x のある範囲は k[l], k[l + 1] , ... , k[r - 1]
    l = 0
    r = n * n

    # 分割できなくなるまで二分探索
    while r - l >= 1:
        i = int((l + r) / 2)
        if kk[i] == x:
            return True
        elif kk[i] < x:
            l = i + 1
        else:
            r = i

    # 見つからなければ False
    return False


# 条件を満たす組み合わせがあるか？
f = False

# ブルートフォース
for a in range(n):
    for b in range(n):
        if binary_search(m - k[a] - k[b]):
            f = True

if (f):
    print("YES!")
else:
    print("NO...")

File: ch01/cpp/lottery_fastest.cpp
#include <cstdio>
#include <algorithm>

const int MAX_N = 1000;

// 入力
int n, m, k[MAX_N];
// Kc と Kd の和を格納しておく配列
int kk[MAX_N * MAX_N];

bool binary_search(int x) {
  // x の存在しうる範囲は kk[l], kk[l + 1], ... , kk[r - 1]
  int l = 0, r = n;

  // 上記の範囲に何も含まれるなるまで繰り返す
  while (r - l >= 1) {
    int i = (l + r) / 2;
    if (kk[i] == x) return true;
    else if (kk[i] < x) l = i + 1;
    else r = i;
  }

  // 見つからなければ false を返す
  return false;
}

void solve() {
  // k[c] + k[d] の取りうる値を計算する
  for (int c = 0; c < n; c++) {
    for (int d = 0; d < n; d++) {
      kk[c * n + d] = k[c] + k[d];
    }
  }

  // ソートする
  std::sort(kk, kk + n * n);

  // 和が m になる組み合わせがあるか？
  bool f = false;

  // ブルートフォース
  for (int a = 0; a < n; a++) {
    for (int b = 0; b < n; b++) {
      // 二分探索
      if (binary_search(m - k[a] - k[b])) {
        f = true;
      }
    }
  }

  // 結果を出力
  if (f) puts("YES!");
  else puts("NO...");
}

int main() {
  scanf("%d %d", &n, &m);
  for (int i = 0; i < n; i++) {
    scanf("%d", &k[i]);
  }

  solve();

  return 0;
}
